src/backend/columnar/cstore_metadata_tables.c

```diff
@@ -1097,55 +1042,149 @@ ColumnarOptionsIndexRegclass(void)
 }
 
 
 /*
- * CStoreSkipNodesRelationId returns relation id of cstore_skipnodes.
- * TODO: should we cache this similar to citus?
+ * CStoreNamespaceId returns namespace id of the schema we store cstore
+ * related tables.
  */
 static Oid
-CStoreSkipNodesRelationId(void)
+CStoreNamespaceId(void)
 {
-	return get_relname_relid("cstore_skipnodes", CStoreNamespaceId());
+	return get_namespace_oid("cstore", false);
 }
 
 
 /*
- * CStoreSkipNodesIndexRelationId returns relation id of cstore_skipnodes_pkey.
- * TODO: should we cache this similar to citus?
+ * ReadMetapage reads metapage for the given relfilenode. It returns
+ * false if the relation doesn't have a meta page yet.
  */
-static Oid
-CStoreSkipNodesIndexRelationId(void)
+static ColumnarMetapage *
+ReadMetapage(RelFileNode relfilenode, bool missingOk)
 {
-	return get_relname_relid("cstore_skipnodes_pkey", CStoreNamespaceId());
+	StringInfo metapageBuffer = NULL;
+	Oid relationId = RelidByRelfilenode(relfilenode.spcNode,
+										relfilenode.relNode);
+	if (OidIsValid(relationId))
+	{
+		Relation relation = relation_open(relationId, NoLock);
+
+		RelationOpenSmgr(relation);
+		int nblocks = smgrnblocks(relation->rd_smgr, MAIN_FORKNUM);
+		RelationCloseSmgr(relation);
+
+		if (nblocks != 0)
+		{
+			metapageBuffer = ReadFromSmgr(relation, 0, sizeof(ColumnarMetapage));
+		}
+
+		relation_close(relation, NoLock);
+	}
+
+	if (metapageBuffer == NULL)
+	{
+		if (!missingOk)
+		{
+			elog(ERROR, "columnar metapage was not found");
+		}
+
+		return NULL;
+	}
+
+	ColumnarMetapage *metapage = palloc0(sizeof(ColumnarMetapage));
+	memcpy_s((void *) metapage, sizeof(ColumnarMetapage),
+			 metapageBuffer->data, sizeof(ColumnarMetapage));
+
+	return metapage;
 }
 
 
 /*
- * CStoreNamespaceId returns namespace id of the schema we store cstore
- * related tables.
+ * InitMetapage initializes metapage for the given relation.
  */
-static Oid
-CStoreNamespaceId(void)
+static ColumnarMetapage *
+InitMetapage(Relation relation)
 {
-	return get_namespace_oid("cstore", false);
+	ColumnarMetapage *metapage = palloc0(sizeof(ColumnarMetapage));
+	metapage->storageId = GetNextStorageId();
+	metapage->versionMajor = CSTORE_VERSION_MAJOR;
+	metapage->versionMinor = CSTORE_VERSION_MINOR;
+
+	/* create the first block */
+	Buffer newBuffer = ReadBuffer(relation, P_NEW);
+	ReleaseBuffer(newBuffer);
+
+	Assert(sizeof(ColumnarMetapage) <= BLCKSZ - SizeOfPageHeaderData);
+	WriteToSmgr(relation, 0, (char *) metapage, sizeof(ColumnarMetapage));
+
+	return metapage;
+}
+
+
+/*
+ * GetNextStorageId returns the next value from the storage id sequence.
+ */
+static uint64
+GetNextStorageId(void)
+{
+	Oid sequenceId = get_relname_relid("cstore_storageid_seq", CStoreNamespaceId());
+	Datum sequenceIdDatum = ObjectIdGetDatum(sequenceId);
+
+	/*
+	 * Generate new and unique storage id from sequence.
+	 * TODO: should we restrict access to the sequence, which might require
+	 * switching security context here?
```

Comment:
What can go wrong if this sequence is not protected? I feel we should conclude this TODO before merging.

The only thing I can see going wrong here is that the user decreases the sequence causing the same storage id to be reused. Is there any way we can enforce uniqueness of the storage id when fetching one from the sequence?

When the user decreases the sequence I'd rather throw an error next time we try to create a table saying the storage id was already in use. Then we don't need any restrictions, and it will also guard against a wrap around (which, even on 32 bit, I deem unlikely here, let alone 64 bit we return).

Reply:
We need to do a security context switch anyway, since `nextval` does ACL checks and without the security context switch restricted users won't be able to create columnar tables. I'll add some multiuser tests in a separate PR.

Jeff's unique identifier generation is an interesting one. I kept it for consideration in future PRs.